feat: Support for Platform "Gerrit" (#18961)
renovatebot/renovate: docs/usage/configuration-options.md

For example: you're following the `next` tag, but later the stream you actually want is called `stable` instead.
If `next` is no longer getting updates, you must switch your `followTag` to `stable` to get updates again.

## gerritAutoApprove

> If set to `true`, Renovate will automatically approve (assign Code-Review label = +2) their own changes.

Reviewer: ```suggestion
If set to `true`, Renovate automatically approves its own changes.
This means Renovate assigns the Code-Review label = +2.
```

Only the `onboardingConfigFileName` (which defaults to `renovate.json`) is supported for `forkProcessing`.
You can't use other filenames because Renovate only checks the default filename when using the Git-hosting platform's API.

## gerritLabelMapping

The name of the Gerrit labels to use for the `StabilityDays` and `Merge-Confidence` status.

```json
{
  "stabilityDaysLabel": "Renovate-Stability",
  "mergeConfidenceLabel": "Renovate-Confidence"
}

> ```

Reviewer: You probably forgot to update this part. 😉 

We renamed `stabilityDays` to `minimumReleaseAge`.

Author: sorry

> If one or both are unset, then these information will not be passed to the Gerrit changes.

Reviewer: ```suggestion
You must set _both_ labels, for Renovate to pass the information to the Gerrit changes.
```

I think this is easier to understand, check if it's technically correct. 😉

Author: no,both features can be used independently.

I'm still unsure if the `minimumReleaseAge` is really useful, because if the release-age is not reached then no PR will be created and therefore no branch-status will be set.

The status can't have any other value then green or?

Reviewer: > no,both features can be used independently.

How about this then:

```suggestion
You must set a label, for Renovate to pass the information to the Gerrit changes.
```

> I'm still unsure if the `minimumReleaseAge` is really useful, because if the release-age is not reached then no PR will be created and therefore no branch-status will be set.

I think your PR will allow Gerrit-users to set labels for Renovate's PRs? I don't see how setting a `minimumReleaseAge` label is helpful in that case. If the `minimumReleaseAge` is _not_ met, Renovate won't create the PR in the first place. So in effect users would add a `minimumReleaseAge` label that tells them nothing special?

> The status can't have any other value then green or?

I don't know how `renovate/stability-days` works internally. I have seen this behavior myself:

1. Use Dependency Dashboard to force Renovate to create a PR that's still watiting for the `minimumReleaseAge` days to be met
2. Renovate creates PR
3. Look at `renovate/stability-days`, that's _not_ met, see yellow check box
4. Once `minimumReleaseAge` is reached, check turns green

So it looks like there are at least two states, yellow and green. There may be a third red state though, or other states. 😄

Author: Okay, but with Gerrit there is no Dependency-Dashboard to force the creation of PRs.
Then this label seems to be useless for Gerrit. I think i should remove this label.

Is it possible to test the Merge-Confidence Feature? 
Last time I've tried it looks like i need a special subscription (Token?) for this feature?

Author: > > Okay, but with Gerrit there is no Dependency-Dashboard to force the creation of PRs. Then this label seems to be useless for Gerrit. I think i should remove this label.
> 
> No, It depends on `internalChecksFilter`. https://docs.renovatebot.com/configuration-options/#minimumreleaseage
> 

Okay, good to know. But i've already removed the Gerrit-Support for this label.

And I guess that setting `internalChecksFilter=none` will probably only affect PR creation but not the creation of branches?

And with Gerrit each branch creation will automatically create a PR. (implicite)

Reviewer: ```suggestion
You must set at least _one_ of the labels above for Renovate to pass the information to the Gerrit changes.
```

I think the only thing left to do here is to reword the sentence?

Let's put our points positively.

Only the `onboardingConfigFileName` (which defaults to `renovate.json`) is supported for `forkProcessing`.
You can't use other filenames because Renovate only checks the default filename when using the Git-hosting platform's API.

> ## gerritLabelMapping

Reviewer 2: I would like to supersede this option with #23530 which will be more generic

Reviewer: @rarkins what's the next step to get this PR moving again?

Reviewer 2: https://github.com/renovatebot/renovate/pull/26047 was merged, so this `gerritLabelMapping` should be replaced with a guidance on how to use `statusCheckNames` instead

@sselberg maybe you can contribute this change?

Author: @rarkins as far as i understand the new `statusCheckNames` the setBranchStatus/getBranchStatus for each Platform will only be called if i've configured the correspinding statusCheckName for one of the checks?

Therefore i can remove my own `gerritLabelMapping` config and directly use the`context` Parameter from set/getBranchStatus as an Gerrit-Label?

Reviewer 2: @NiasSt90 yes that's right. Please use statusCheckNames just like you were planning to use gerritLabelMapping. I guess you would want to populate default values if platform=gerrit?

Author: @rarkins  it's done. 

And i don't think we need default values for Gerrit here. 
Because `minimumReleaseAge` is not needed in most cases, because the default behavior of renovate delays the change-creation until `stabilityDays` are reached and `mergeConfidence` is a paid feature as far as i know (api key required?).
